tmt/steps/provision/__init__.py

```diff
@@ -3482,3 +3511,183 @@ def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[Act
+
+        :param plan: the plan containing provisioned guests and configuration.
+        """
+        self._logger = plan._logger
+        self._plan = plan
+
+    def _load_layout(self, layout_path: Optional[str] = None) -> dict[str, Any]:
+        """
+        Load inventory layout from file or use default.
+
+        :param layout_path: path to a custom layout file, relative to the FMF file.
+        # TODO check layout path logic
+        :returns: dictionary representing the inventory layout structure.
+        """
+        if layout_path:
+            try:
+                # Resolve path relative to the directory containing the FMF file
+                if self._plan.node.sources:
+                    # Use the directory of the first source (where the FMF file is)
+                    fmf_dir = Path(self._plan.node.sources[0]).parent
+                    resolved_path = fmf_dir / layout_path
+                else:
+                    # Fallback to current working directory if no sources
+                    resolved_path = Path(layout_path)
+                return tmt.utils.yaml_to_dict(resolved_path)
+            except tmt.utils.GeneralError as exc:
+                self._logger.warning(f"Failed to load inventory layout: {exc}")
+                return self._default_layout()
+        return self._default_layout()
+
+    def _default_layout(self) -> dict[str, Any]:
+        """
+        Create default inventory layout.
+
+        :returns: basic inventory structure with 'all' and 'ungrouped' groups.
+        """
+        return {'all': {'children': {'ungrouped': {}}}}
+
+    def _get_guest_ansible_config(self, guest: Guest) -> Optional[dict[str, Any]]:
+        """
+        Get ansible configuration for a specific guest from provision step data.
+
+        :param guest: the guest to get configuration for.
+        :returns: ansible configuration dictionary if found, ``None`` otherwise.
+        """
+        # Find the provision step data for this guest
+        provision_step = self._plan.provision
+        for phase in provision_step.phases(classes=ProvisionPlugin):
+            if phase.guest and phase.guest.name == guest.name:
+                return getattr(phase.data, 'ansible', None)
+        return None
+
+    def _get_host_vars(self, guest: Guest) -> dict[str, Any]:
+        """
+        Get host variables for a guest.
+
+        :param guest: the guest to get variables for.
+        :returns: dictionary of Ansible host variables.
+        """
+        vars_dict: dict[str, Any] = {  # TODO check vars
+            'ansible_host': guest.primary_address,
+            'ansible_user': getattr(guest, 'user', None),
+            'ansible_port': getattr(guest, 'port', None),
+        }
+
+        # Add any custom variables from provision step data's ansible.vars
+        ansible_config = self._get_guest_ansible_config(guest)
+        if ansible_config and 'vars' in ansible_config:
+            vars_dict.update(ansible_config['vars'])
```

Comment:
This will become a simple `vars.update(guest.ansible.vars)` - `guest.ansible` will be a container, `guest.ansible.vars` will always be a dictionary, sometimes empty, true, but always a dictionary, no need for doubt and double-checking. We check once, on the boundary, when the raw user data enter tmt digestive tract :)